Station hierarchy 4

irees opened the pull request:
WIP. A fresh attempt at Station Hierarchies, #290.

Resolves #256 
Resolves #257

irees commented:
Cleanup tasks.
1. General code cleanup before review
2. Aggregate operator_serving_stop/route_serving_stop at top level station in stop_stations endpoint
3. Make connected_osm_way_id Stop attribute
4. Transfer type is enumerable
5. Don't override existing parent_stop_onestop_id
6. Create StopEgresses when location_type=2
7. Discuss with routing

irees commented:
TODO: dedup aggreaged operator_serving_stop  / route_serving_stop

drewda commented on app/controllers/api/v1/stop_stations_controller.rb:
```diff
@@ -0,0 +1,33 @@
+#  index_current_stops_on_updated_at      (updated_at)
+#
+
+class Api::V1::StopStationsController < Api::V1::StopsController
```
I imagine the stop_stations endpoint will need different/more query params than the stops endpoint. Say, querying for the stop_stations that include a certain platform.

Not necessary to squeeze into this initial PR, but let's talk more in our call tomorrow with @gknisely and other about the range of possible query parameters for filtering and sorting.

drewda commented on app/models/json_schemas/stop.json:
```diff
@@ -13,6 +13,9 @@
     "tags": { "type": "object" },
     "name": { "type": "string" },
+    "parentStopOnestopId": {
+      "type": ["string", "null"]
```
Is it possible to specify format as the stop Onestop ID regex validator? Or is that not compatible with the combo of string-or-null?

drewda commented on app/models/stop.rb:
```diff
@@ -71,23 +77,20 @@ def csv_row_values
-      },
-      changeset: changeset
-    )
+    before_update_making_history(changeset)
```
How about a more descriptive name for this function call? More below...

drewda commented on app/models/stop.rb:
```diff
@@ -71,23 +77,20 @@ def csv_row_values
   end
   def before_update_making_history(changeset)
+    super(changeset)
     OperatorRouteStopRelationship.manage_multiple(
```
Lots going on here. How about separating 103-106 and 107-133 into two separate private methods with descriptive names? Then you can call them in sequence from both `after_create_making_history` and `before_update_making_history`.

drewda commented on app/models/stop.rb:
```diff
@@ -284,13 +322,60 @@ def clean_attributes
 end
 
+class StopPlatform < Stop
+  # validates :parent_stop, presence: true
```
Uncomment?

drewda commented on app/serializers/stop_station_serializer.rb:
```diff
@@ -0,0 +1,107 @@
+  # Aggregate operators_serving_stop
+  def operators_serving_stop
+    result = object.operators_serving_stop
+    object.stop_platforms.each { |sp| result |= sp.operators_serving_stop}
```
This will load many models in to memory. Better to write as a chained ActiveRecord query, and put in a method on the stop model called, say, `operators_serving_stop_and_platforms`.

Not a blocker to merging in this PR, but should happen before launch.

drewda commented on app/serializers/stop_station_serializer.rb:
```diff
@@ -0,0 +1,107 @@
+  # Aggregate routes_serving_stop
+  def routes_serving_stop
+    result = object.routes_serving_stop
+    object.stop_platforms.each { |sp| result |= sp.routes_serving_stop}
```
Again, better to ues a full DB query and to move to a method on the model (that will be better named and testable).

drewda commented on app/services/onestop_id.rb:
```diff
@@ -182,16 +208,10 @@ def validate_hash(value)
-      num_components = split.size
-      LOOKUP[[prefix, num_components]]
-    end
+    OnestopId::OnestopIdBase.descendants.select { |cls| cls.match?(string) }.first
```
Any idea on the performance implications of this change? I imagine it's called many times.

drewda commented on db/migrate/20160526224318_create_stop_transfers.rb:
```diff
@@ -0,0 +1,18 @@
+        t.hstore :tags
+        t.references :stop, index: true
+        t.references :to_stop, class_name: "Stop", index: true
+        t.references :created_or_updated_in_changeset #, index: true # TODO: short index name
```
TODO?

drewda commented on debug_gtfs_graph.rb:
```diff
@@ -0,0 +1,39 @@
+ActiveRecord::Base.logger = Logger.new(STDOUT)
```
Is there a better place to put this than the root? Or is it a temporary test script that's better not to commit to the repo?

drewda commented on spec/services/gtfs_graph_spec.rb:
```diff
@@ -68,7 +70,7 @@
           "gtfs://f-9q9-caltrain/s/ctsj"
         )
-        expect(s.timezone).to eq('America/Los_Angeles')
+        # expect(s.timezone).to eq('America/Los_Angeles')
```
Why the comment? Change in Caltrain feed?

drewda commented:
@irees This is looking good and quite close to being ready for merge into master and deploy to dev.

I put a few comments, questions, and suggestions inline. Please have a look. Some are not blockers to this being merged in. Your choice which to get in with this PR vs. which to create individual follow-up issues for.

irees commented:
@drewda I've addressed a few of your comments -

1. Uncommented debugging items, removed debug script
2. Stop json schema 'parentStopOnestopId' now uses 'stop-onestop-id' format
3. Refactored Stop 'before_update_making_history' method into smaller private methods
4. Checked and fixed slowness on OnestopId::handler_by_string

I've left the StopStationController as a simple subclass of StopController for now; it can be updated with feedback from Greg.

I'll also implement the suggestion to create a 'routes_serving_stop_and_platforms' method. There are a few details on how to handle removing duplicates that I'll write up in a separate comment.

irees commented:
Full StopStationsController

irees commented:
- [ ] Test with feed transitions

drewda commented on app/models/stop.rb:
```diff
@@ -116,12 +115,39 @@ def before_destroy_making_history(changeset, old_model)
+      .where('stop_id IN (?) OR stop_id = ?', Stop.where(parent_stop_id: self.id).select(:id), self.id)
+      .select(:operator_id)
+      .uniq
+      .map { |o| OperatorServingStop.new(stop: self, operator_id: o.operator_id) }
```
@irees Would the following work to return the same results, using only the DB:
```ruby
    OperatorServingStop
      .where('stop_id IN (?) OR stop_id = ?', Stop.where(parent_stop_id: self.id).select(:id), self.id)
      .select('DISTINCT ON (current_operators_serving_stop.operator_id) *')
```
?

drewda commented on spec/factories/stop_factory.rb:
```diff
@@ -29,6 +29,8 @@
 #  index_current_stops_on_updated_at      (updated_at)
 #
 
+Stop.connection
```
What does this line do?

drewda commented:
@irees only one horribly small question on my end: why the `Stop.connection` line in the stop spec? Otherwise, this looks ready to merge/deploy to dev after that small prod release goes out.

irees commented:
@drewda The `Stop.connection` line was a workaround to rails autoloader not loading the STI models contained in `stops.rb`. I broke StopPlatform, StopEgress into separate files for model, specs, factories, etc., and now it works fine.
